### PR TITLE
Handle common disk failures in the spill code path

### PR DESCRIPTION
**Pull request by dfinninger**
In a JBOD setup disk failures are quite common. In the absence of basic failure
handling, a presto worker will not recover until a config is changed and process
is restarted -- which is difficult to handle operationally.

Spill path discovery routine already handles the case when one or more paths
are out of free space, this change adds couple of additional checks to cover
out of band disk failures.

### CODE REVIEW

**Comment by kokosing** [comment]
How does it differ to https://github.com/prestosql/presto/pull/2175?

**Comment by kokosing** [comment]
> How does it differ to #2175?

Please nevermind, I just noticed the comment: https://github.com/prestosql/presto/pull/2175#issuecomment-572324028

**Comment by kokosing** [comment]
@dfinninger Please sing CLA: https://github.com/prestosql/cla

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java [hunk 1]
```diff
@@ -60,6 +69,7 @@
     @VisibleForTesting
     static final String SPILL_FILE_SUFFIX = ".bin";
     private static final String SPILL_FILE_GLOB = "spill*.bin";
+    private static final Duration SPILL_PATH_HEALTH_EXPIRY_INTERVAL = Duration.ofSeconds(300);
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java]: Please use `ofMinutes(5)`

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java [hunk 2]
```diff
@@ -106,13 +117,18 @@ public FileSingleStreamSpillerFactory(
...
-                throw new IllegalArgumentException(format("spill path %s is not writable; adjust %s config property or filesystem permissions", path, SPILLER_SPILL_PATH));
+            if (!isAccessible(path)) {
+                throw new IllegalArgumentException(format("spill path %s is not accessible, it must be +rwx; adjust %s config property or filesystem permissions", path, SPILLER_SPILL_PATH));
             }
         });
         this.maxUsedSpaceThreshold = maxUsedSpaceThreshold;
         this.spillEncryptionEnabled = spillEncryptionEnabled;
         this.roundRobinIndex = 0;
+
+        CacheLoader<Path, Boolean> cacheLoader = CacheLoader.from(path -> isAccessible(path) && isSeeminglyHealthy(path));
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java]: please inline `cacheLoader`

**Comment by kokosing** [comment]
@shrijeet, @dfinninger have you been using this patch in production?

**Comment by electrum** [comment]
As a follow up, we should proactively run the check (or clear the cache to force a check) anytime we get an IO error while reading or writing in the spill code.

**Comment by dfinninger** [comment]
Ha, great time for my printer to be out of ink. I'll need to pick up some replacement cartridges and I'll get the CLA signed and back to y'all by tomorrow.

**Comment by dfinninger** [comment]
@kokosing I have not used this in production, no.

**Comment by dfinninger** [comment]
Alright, just submitted the CLA.

**Comment by kokosing** [comment]
@cla-bot check

**Comment by kokosing** [comment]
> As a follow up, we should proactively run the check (or clear the cache to force a check) anytime we get an IO error while reading or writing in the spill code.

@dfinninger are you going to handle that?

**Comment by dfinninger** [comment]
@kokosing I could take a look at it this week. And I'm assuming you mean in a follow-up PR?

**Comment by kokosing** [comment]
>  And I'm assuming you mean in a follow-up PR?

Yes.

**Comment by kokosing** [comment]
Merged, thanks!

**Pull request merged**